Fix broken tests in Node v6.5 and update supported node versions

meeber opened the pull request:
- #783 is a prerequisite to this PR
- The first commit fixes a number of tests that were broken in Node v6.5 due to a change in how anonymous functions that are assigned to variables appear in failed assertion messages (from "[Function]" to "[Function: _var_name_]". 
- The second commit updates Travis Node versions (adds all versions of node that are officially maintained: 0.10, 0.12, 4, 6; also adds safety net aliases: lts/*, node).
- The third commit updates package.json's engines field to node v0.10 since it's the minimum version that we officially support.

Notes on Travis node versions:
- If there was only ever a total of two versions actively supported at the same time, one LTS and one Current, then the lts/* and node aliases would alone be sufficient. The challenge comes from the planned overlaps that are built into the Node release schedule, causing there to periodically be two LTS or two Current versions actively supported at once. The only way to prevent an actively supported version from being prematurely dropped from automated testing during these periods of overlap is to manually list all actively supported versions. The aliases lts/* and node should still be included as safety nets, ensuring that there's still good (but not perfect) coverage if package maintainers fail to keep on top of manually updating the version list. A downside of providing this safety net is that during periods of non-overlap, the test suite will run twice for the same release.

- Whether or not to use numeric versions like 4 or aliased versions like lts/argon is merely preference. I prefer the former because I think it more accurately reflects how most people discuss and interact with different Node versions.

- Including expiration date comments for each version in .travis.yml is good.

keithamus commented:
These changes LGTM. We might want to consider dropping Node 0.10 for a clean break with 4.0 of chai. Should probably still support 0.12 for now though

meeber commented:
@keithamus I'd definitely like to discuss strategy for Node legacy support sometime soon. I haven't solidified my opinion on the issue yet. Thinking about it casually, I'm torn multiple ways.

keithamus commented:
@meeber I was going to write it here, but instead I decided to move my comment to #124. Let's continue discussion over there, and try to ratify something that users can read too.

meeber commented:
@keithamus I read your comment in the other thread and agree with you completely. I've updated this PR so that .travis.yml explicitly states all node versions that are officially maintained. I also added a third commit to bump up the minimum node version in package.json's engines field.

lucasfcosta commented:
LGTM.
I totally agree with @keithamus comment on #124.
Also, nice job adding the expire dates of the engines we're testing on `.travis.yml` file.

meeber commented:
@lucasfcosta Thanks for reviewing! Adding expire dates in `.travis.yml` was @keithamus's idea :D

I'm not sure we've reached complete consensus on the version list per https://github.com/chaijs/type-detect/pull/69. @keithamus?

keithamus commented:
Yup I think this is totally right and sets the precedent for all other projects in the chaijs repo. Awesome work, as usual @meeber 😄
